update: add aufs to overlay migration

robertgzr opened the pull request:
This adds the AUFS to OverlayFS migration utility [(to be) built into balenaEngine](https://github.com/balena-os/balena-engine/pull/168) to the update hooks.

I tested this on rpi3 and fin so far and it worked fine :tada: 

---
### Contributor checklist
<!-- For completed items, change [ ] to [x].  -->
- [x] Changes have been tested
- [ ] `Change-type` present on at least one commit
- [ ] `Signed-off-by` is present
- [ ] The PR complies with the [Open Embedded Commit Patch Message Guidelines](http://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines)
<!-- optional: `Changelog-entry` present on at least one commit if you want to set the changelog entry manually-->

resin-jenkins commented:
Can one of the admins verify this patch?

floion commented:
@resin-jenkins ok to test

ZubairLK commented on meta-balena-common/conf/distro/include/balena-os.inc:
```diff
@@ -67,8 +67,8 @@ INITRAMFS_TASK = ""
-# Default the docker storage driver to aufs
-BALENA_STORAGE ?= "aufs"
+# Default the docker storage driver to overlay2
+BALENA_STORAGE ?= "overlay2"
```
need to check this PR works with devices that specify aufs (due to old kernels etc)

robertgzr replied:
What devices would that be? Intel Quark, ...? I'm not sure I have one test that myself here

roman-mazur replied:
Rpi will have this support, won't it?
I'm looking forward to this change so I can build balenaOS on my Mac for my RaspberryPi device :)

ZubairLK replied:
@robertgzr one of the odroids I think.

robertgzr replied:
We still need to verify this I think... does anybody from the devices team have the devices relevant to this?

cc @floion @horia-delicoti

ZubairLK commented on meta-balena-common/recipes-core/balena-rollback/files/rollback-tests:
```diff
@@ -36,6 +36,10 @@ fi
 if /usr/lib/balena/balena-healthcheck ; then
 	echo "Rollback: Balena looks healthy"
+	echo "Rollback: Cleaning up storage migration..."
+	env BALENA_A2O_STORAGE_ROOT="/mnt/data/docker" a2o-migrate -debug -cleanup
```
I see cleanup. But not really a test if overlay is working?
Also, checked a2o-migrate cleanup, I'd recommend there should be a basic sanity check if the overlay directory is present or not.

ZubairLK commented on meta-balena-common/recipes-support/hostapp-update-hooks/files/60-balena_migrate_storage:
```diff
@@ -0,0 +1,33 @@
+
+DURING_UPDATE=${DURING_UPDATE:-0}
+
+# TODO detect if we are actually going to boot into overlay and only migrate if that's the case
```
Still todo?

robertgzr replied:
https://github.com/balena-os/meta-balena/pull/1577#discussion_r300400432 ?

robertgzr commented:
- changed base to `development`
- renamed hooks scripts to match hooks v2 pr
- this probably conflicts with #1679 since I rebased my engine pr on top of latest master too...

robertgzr commented:
@ZubairLK with hooks v2 I still need a check that would trigger a rollback in case it fails, right?

robertgzr commented:
I guess the jenkins failure is expected since this PR is based on `development` which doesn't have the hooks v2 work from #1595

ZubairLK commented on meta-balena-common/recipes-support/hostapp-update-hooks/hostapp-update-hooks.bb:
```diff
@@ -9,6 +9,7 @@ inherit allarch
 
 HOSTAPP_HOOKS = " \
     0-bootfiles \
+    60-balena_migrate_storage \
```
need to add all the files and the directory name separately. See https://github.com/balena-os/meta-balena/commit/962e526b2582d66ea4ad15433a45445e1ee9327f#diff-504d12ea9d623c9bcfafcd9e84037198 for an example.

ZubairLK commented on meta-balena-common/recipes-containers/balena/balena_git.bb:
```diff
@@ -11,10 +11,10 @@ LIC_FILES_CHKSUM = "file://src/import/LICENSE;md5=9740d093a080530b5c5c6573df9af4
-BALENA_VERSION = "18.09.8-dev"
-BALENA_BRANCH= "master"
+BALENA_VERSION = "18.9.10-dev"
+BALENA_BRANCH= "a2o-migrate"
```
I'm afraid, I'd like to avoid switching the branch. Any reason why a2o-migrate isn't in master?

robertgzr replied:
this is not permanent. this pr is still a draft and so is https://github.com/balena-os/balena-engine/pull/168 because I still haven't actually tested this with the hoooks v2 stuff

ZubairLK commented on meta-balena-common/recipes-support/hostapp-update-hooks/files/60-balena_migrate_storage/forward:
```diff
@@ -0,0 +1,15 @@
+env BALENA_A2O_STORAGE_ROOT=$USER_APP_STORAGE a2o-migrate -debug -migrate
+
+printf "[INFO] balena_migrate_storage: Migrating hostapp storage to overlayfs... \n"
+env BALENA_A2O_STORAGE_ROOT=$HOST_APP_STORAGE a2o-migrate -debug -migrate
```
Not sure I understand this. Shouldn't this be left for the device to decide?
We fetch an aufs image. And then on-device it becomes overlay? feels wrong.

imrehg commented:
@balena-ci retest please?

robertgzr commented:
according to https://github.com/balena-os/balena-engine/pull/168#issuecomment-617235123 we decided to ditch the approach making use of HUP hooks. 

A future version of the engine will take care of the migration if overlayfs is enabled by the OS and we will fall back to rollback if that is unsuccessfull.
